be/src/storage/rowset/data_sample.cpp

```diff
@@ -0,0 +1,342 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/rowset/data_sample.h"
+
+#include <fmt/format.h>
+
+#include <random>
+#include <stdexcept>
+
+#include "storage/olap_common.h"
+#include "storage/types.h"
+#include "storage/zone_map_detail.h"
+#include "types/logical_type.h"
+#include "util/runtime_profile.h"
+
+namespace starrocks {
+
+StatusOr<RowIdSparseRange> BlockDataSample::sample(OlapReaderStatistics* stats) {
+    RETURN_IF(_probability_percent == 0 || _probability_percent == 100,
+              Status::InvalidArgument("percent should be in (0, 100)"));
+    std::mt19937 mt(_random_seed);
+    std::bernoulli_distribution dist(_probability_percent / 100.0);
+
+    size_t sampled_blocks = 0;
+    size_t total_blocks = _total_rows / _rows_per_block;
+    RowIdSparseRange sampled_ranges;
+    for (size_t i = 0; i < _total_rows; i += _rows_per_block) {
+        if (dist(mt)) {
+            sampled_blocks++;
+            sampled_ranges.add(RowIdRange(i, std::min(i + _rows_per_block, _total_rows)));
```

Comment:
- can RowIdRange span two blocks ? a interesting question is what's the math expect of number of RowIdRanges spanning two block when stride(_rows_per_block) varies. is there an optimized stride?

Reply:
i would say not, each RowIdRange is expected to cover exactly one block